catalyst/dl/callbacks/base.py

```diff
@@ -268,26 +270,43 @@ def __init__(
         self.mode = mode
         self.reduce_metric = reduce_metric
 
-    def step(self, state):
+        self._onecycle_recalculated = False
+
+    def step(self, state: RunnerState):
         scheduler = state.get_key(
             key="scheduler", inner_key=self.scheduler_key
         )
 
-        lr, momentum = scheduler_step(
+        valid_metric = \
+            safitty.get(state.metrics.valid_values, self.reduce_metric)
+        lr, momentum = self._scheduler_step(
             scheduler=scheduler,
-            valid_metric=state.metrics.valid_values.get(
-                self.reduce_metric, None)
+            valid_metric=valid_metric
         )
 
         state.set_key(lr, key="lr", inner_key=self.scheduler_key)
         state.set_key(momentum, key="momentum", inner_key=self.scheduler_key)
 
-    def on_stage_start(self, state):
+    def on_stage_start(self, state: RunnerState):
         scheduler = state.get_key(
             key="scheduler", inner_key=self.scheduler_key
         )
         assert scheduler is not None
 
+        if isinstance(scheduler, OneCycleLR) and self.mode == "batch":
+            self._onecycle_recalculated = False
+            scheduler.reset()
+
+    def on_loader_start(self, state: RunnerState):
+        scheduler = state.get_key(
+            key="scheduler", inner_key=self.scheduler_key
+        )
+        if state.loader_name == "train" and \
+                isinstance(scheduler, OneCycleLR) and self.mode == "batch":
+            if not self._onecycle_recalculated:
+                scheduler.recalculate_(loader_len=state.loader_len)
+                self._onecycle_recalculated = True
```

Comment:
looks like it can work incorrectly with more than 1 train loaders, could you please check it with several of them?